### PR TITLE
feat(*): Support string length validation

### PR DESCRIPTION
**Pull request by ragi-dayananda**
# Closes 

WIP

### Related Issues
- Issue #[643](https://github.com/accordproject/concerto/issues/643)
- Pull Request #[concerto-metamodel/pull/3](https://github.com/accordproject/concerto-metamodel/pull/3)

### Author Checklist
- [ ] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [ ] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `main` from `fork:branchname`

### CODE REVIEW

**Comment by dselman** [comment]
Looks great so far @ragi-dayananda! Once the introspection class is in place for validation etc. it would be great to update the JSON Schema code generator to use this as an example / test of it being applied.

**Comment by dselman** [comment]
Docs page to be updated is here: https://concerto.accordproject.org/docs/design/specification/model-properties

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 1]
```diff
@@ -37,17 +38,45 @@ class StringValidator extends Validator{
...
+                this.reportError(null, 'Invalid string length, minLength and-or maxLength must be specified.');
+            } else if (this.minLength < 0 || this.maxLength < 0) {
+                this.reportError(null, 'minLength and-or maxLength must be positive integers.');
+            } else if (this.minLength === null || this.maxLength === null) {
+                // this is fine and means that we don't need to check whether minLength > maxLength
+            } else {
+                if(this.minLength > this.maxLength) {
+                    this.reportError(null, 'minLength must be less than or equal to maxLength.');
+                }
+            }
```
Comment (mttrbrts) [inline packages/concerto-core/lib/introspect/stringvalidator.js:69]:
```suggestion
            } else if(this.minLength > this.maxLength) {
                this.reportError(null, 'minLength must be less than or equal to maxLength.');
            }
```

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 2]
```diff
@@ -37,17 +38,45 @@ class StringValidator extends Validator{
...
+        if (lengthValidator) {
+            if(Object.prototype.hasOwnProperty.call(lengthValidator, 'minLength')) {
+                this.minLength = lengthValidator.minLength;
+            }
+            if(Object.prototype.hasOwnProperty.call(lengthValidator, 'maxLength')) {
+                this.maxLength = lengthValidator.maxLength;
+            }
+            if(this.minLength === null && this.maxLength === null) {
+                // can't specify no upper and lower value
+                this.reportError(null, 'Invalid string length, minLength and-or maxLength must be specified.');
```
Comment (mttrbrts) [inline packages/concerto-core/lib/introspect/stringvalidator.js:60]: Should these be `this.reportError(field.getName(), ...`?

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 3]
```diff
@@ -60,14 +89,38 @@ class StringValidator extends Validator{
...
     validate(identifier, value) {
         if(value !== null) {
-            if(!this.regex.test(value)) {
-                this.reportError(identifier, 'Value \'' + value + '\' failed to match validation regex: ' + this.regex);
+            if (this.regex) {
+                if(!this.regex.test(value)) {
+                    this.reportError(identifier, 'Value \'' + value + '\' failed to match validation regex: ' + this.regex);
+                }
+            }
+            //Enforce string length rule after regex check
```
Comment (mttrbrts) [inline packages/concerto-core/lib/introspect/stringvalidator.js:97]: This comment suggests that length checking should come first.
https://github.com/accordproject/concerto/issues/643#issuecomment-1542134421

File: packages/concerto-core/test/introspect/stringvalidator.js [hunk 4]
```diff
@@ -36,22 +48,51 @@ describe('StringValidator', () => {
...
+                new StringValidator(mockField, null, NEGETIVE_LENGTH);
+            }).should.throw(/minLength and-or maxLength must be positive integers/);
+        });
     });
 
     describe('#validate', () => {
 
         it('should ignore a null string', () => {
-            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' });
+            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' }, VALID_MIN_LENGTH_AND_MAX_LENGTH_AST);
```
Comment (mttrbrts) [inline packages/concerto-core/test/introspect/stringvalidator.js:89]: Why do we need to change the existing tests? It's important that if we skip this parameter, then we should get the same behaviour.

File: packages/concerto-core/test/introspect/stringvalidator.js [hunk 5]
```diff
@@ -36,22 +48,51 @@ describe('StringValidator', () => {
...
             v.getRegex().toString().should.equal('/^[A-z][A-z][0-9]{7}/');
             v.validate('id', null);
         });
 
         it('should validate a string', () => {
-            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' });
+            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' }, VALID_MIN_LENGTH_AND_MAX_LENGTH_AST);
```
Comment (mttrbrts) [inline packages/concerto-core/test/introspect/stringvalidator.js:95]: Why do we need to change the existing tests? It's important that if we skip this parameter, then we should get the same behaviour.

File: packages/concerto-core/test/introspect/stringvalidator.js [hunk 6]
```diff
@@ -87,48 +128,120 @@ describe('StringValidator', () => {
...
+
     });
 
     describe('#validate with custom RegEx engine', () => {
         const options = {
             regExp: XRegExp
         };
         it('should ignore a null string', () => {
-            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' }, options);
+            let v = new StringValidator(mockField, { pattern: '^[A-z][A-z][0-9]{7}' }, VALID_MIN_LENGTH_AND_MAX_LENGTH_AST, options);
```
Comment (mttrbrts) [inline packages/concerto-core/test/introspect/stringvalidator.js:170]: Why do we need to change the existing tests? It's important that if we skip this parameter, then we should get the same behaviour.

File: packages/concerto-core/test/introspect/stringvalidator.js [hunk 7]
```diff
@@ -87,48 +128,120 @@ describe('StringValidator', () => {
...
             v.getRegex().toString().should.equal('/^[A-z][A-z][0-9]{7}/');
             v.validate('id', null);
         });
 
         it('should validate a string', () => {
-            let v = new StringValidator(mockField, { pattern: '^[\\p{Letter}\\p{Number}]{7}', flags: 'u' }, options);
+            let v = new StringValidator(mockField, { pattern: '^[\\p{Letter}\\p{Number}]{7}', flags: 'u' }, VALID_MIN_LENGTH_AND_MAX_LENGTH_AST, options);
```
Comment (mttrbrts) [inline packages/concerto-core/test/introspect/stringvalidator.js:176]: Why do we need to change the existing tests? It's important that if we skip this parameter, then we should get the same behaviour.

**Review by mttrbrts — CHANGES_REQUESTED** [review]
Good progress, Ragi. I've added some minor comments in-line.

Can you update the issue to track the remaining work, please?

For example:

  - [x] validate           validate JSON against model files
  - [ ]  compile            generate code for a target platform
    - [ ] TypeScript
    - [ ] JSON Schema
    - etc...  
  - [x] parse              parse a cto string to a JSON syntax tree
  - [x] print              print a JSON syntax tree to a cto string
  - [ ] compare            compare two Concerto model files
  - [ ] infer              generate a concerto model from a source schema
  - [ ] generate <mode>    generate a sample JSON object for a concept
  
 We don't necessarily need complete coverage before we release this work, however, we do need to make sure that we fail gracefully / warn, or hide this work behind an environment variable until it's ready.

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 8]
```diff
@@ -60,14 +89,38 @@ class StringValidator extends Validator{
      */
     validate(identifier, value) {
         if(value !== null) {
-            if(!this.regex.test(value)) {
-                this.reportError(identifier, 'Value \'' + value + '\' failed to match validation regex: ' + this.regex);
+            if (this.regex) {
```
Comment (Jonathan-Casey) [inline packages/concerto-core/lib/introspect/stringvalidator.js:92]: I think its better to avoid unnecessary nesting of "if" statements here. You could always merge this to one conditional check which would improve code readability

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 9]
```diff
@@ -60,14 +89,38 @@ class StringValidator extends Validator{
...
+                if(!this.regex.test(value)) {
+                    this.reportError(identifier, 'Value \'' + value + '\' failed to match validation regex: ' + this.regex);
```
Comment (Jonathan-Casey) [inline packages/concerto-core/lib/introspect/stringvalidator.js:94]: why not use string interpolation for this reportError?

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 10]
```diff
@@ -37,17 +38,45 @@ class StringValidator extends Validator{
...
+            if(Object.prototype.hasOwnProperty.call(lengthValidator, 'maxLength')) {
+                this.maxLength = lengthValidator.maxLength;
+            }
+            if(this.minLength === null && this.maxLength === null) {
+                // can't specify no upper and lower value
+                this.reportError(null, 'Invalid string length, minLength and-or maxLength must be specified.');
+            } else if (this.minLength < 0 || this.maxLength < 0) {
+                this.reportError(null, 'minLength and-or maxLength must be positive integers.');
+            } else if (this.minLength === null || this.maxLength === null) {
+                // this is fine and means that we don't need to check whether minLength > maxLength
```
Comment (Jonathan-Casey) [inline packages/concerto-core/lib/introspect/stringvalidator.js:64]: Is this necessary if there is no code executed on the condition evaluating to true?

File: packages/concerto-core/lib/introspect/stringvalidator.js (hunk 9 above)
Comment (ragi-dayananda) [reply packages/concerto-core/lib/introspect/stringvalidator.js:94]: that's an existing code.

File: packages/concerto-core/lib/introspect/stringvalidator.js (hunk 10 above)
Comment (ragi-dayananda) [reply packages/concerto-core/lib/introspect/stringvalidator.js:64]: I copied most of the logic from number validator. I overlooked it may be. Let me remove it

File: packages/concerto-core/lib/introspect/stringvalidator.js (hunk 3 above)
Comment (ragi-dayananda) [reply packages/concerto-core/lib/introspect/stringvalidator.js:97]: I can update the logic, but I don't think order of execution is gonna solve the problem if regex has length constraint in it

Comment (mttrbrts) [reply packages/concerto-core/lib/introspect/stringvalidator.js:97]: You're right, it won't. We just have the design principle to apply the least complex validation first.

File: packages/concerto-core/lib/introspect/stringvalidator.js [hunk 11]
```diff
@@ -37,17 +38,45 @@ class StringValidator extends Validator{
...
+        this.minLength = null;
+        this.maxLength = null;
+        this.regex = null;
+
+        if (lengthValidator) {
+            if(Object.prototype.hasOwnProperty.call(lengthValidator, 'minLength')) {
+                this.minLength = lengthValidator.minLength;
+            }
+            if(Object.prototype.hasOwnProperty.call(lengthValidator, 'maxLength')) {
+                this.maxLength = lengthValidator.maxLength;
```
Comment (dselman) [inline packages/concerto-core/lib/introspect/stringvalidator.js:56]: this.minLength = lengthValidator?.minLength;
this.maxLength = lengthValidator?.maxLength;

**Review by dselman — COMMENTED** [review]
Good work.

File: packages/concerto-core/test/introspect/stringvalidator.js (hunk 7 above)
Comment (ragi-dayananda) [reply packages/concerto-core/test/introspect/stringvalidator.js]: I can't skip that param, for now passing null instead of valid length ast. 
Since here we are passing `options` as one more param, validator will try to bind it to length validator if we don't explicitly pass null/valid length validator

**Comment by mttrbrts** [comment]
@ragi-dayananda There are a few edge cases missing from your tests. Can you bump the coverage, please?
https://coveralls.io/jobs/121606352/source_files/8583783901

**Review by mttrbrts — APPROVED** [review]
